src/libexpr/primops.cc

```diff
@@ -2750,6 +2750,109 @@ static RegisterPrimOp primop_catAttrs({
     .fun = prim_catAttrs,
 });
 
+static void prim_functionStrict(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.forceValue(*args[0], pos);
+    if (args[0]->isPrimOpApp() || args[0]->isPrimOp()) {
+        v.mkNull();
+        return;
+    }
+
+    state.forceFunction(*args[0], pos, "while evaluating the argument passed to builtins.functionStrict");
+    const auto fun = args[0]->lambda.fun;
+
+    if (args[0]->isLambda() && fun->hasFormals()) {
+        v.mkBool(true);
+    } else {
+        v.mkNull();
+    }
+}
+
+static RegisterPrimOp primop_functionStrict({
```

Comment:
This should be `functionHasFormals` or whatever. It is not immediately obvious to the casual user what a “strict function” is supposed to be (and it is a bit ambiguous as well, I'd say). A function with formals is immediately syntactically obvious, so there is no confusion what this builtin would check for.

Additionally, not all functions that are strict in their argument have formals, as the documentation also admits, e.g.:

```nix
foo: builtins.seq foo (/* body … */)
```

Reply:
I couldn't care less about the syntax of a function, which is why I've named these primops after the semantics.

By calling it _has formals_ we can't go back and add a syntax to make functions with formals lazy. What matters is that the function definition is trivially strict, so that's what the name reflects. Should it be `functionTriviallyStrict`?

If this function is too contentious, I might remove it because it turned out that this one wasn't necessary for the use case I'm trying to solve. I think it's good to have, but low impact, so I don't want to waste time on it.


> not all functions that are strict in their argument have formals

This is for improving error messages in a few cases, and nothing more.

Reply:
Well the semantics are fundamentally not introspectable, so the naming is very deceptive.

> we can't go back and add a syntax to make functions with formals lazy

It wouldn't really matter, the introspection would still do its job as advertised—there would be a breaking change in the semantics of the language, but that would be a problem regardless of the existence of that builtin.